posts/inside-rust/2023-11-09-spec-vision.md

```diff
@@ -0,0 +1,206 @@
+
+**"semantics"**: Rust has a static and dynamic semantics.
+The static semantics of Rust dictates which programs are accepted in the language, while the dynamic semantics determines which of those accepted programs are well-defined, as well as their respective meanings.
+The word "semantics" in the purpose statement refers to both the static and dynamic semantics of Rust collectively.
+
+**"current", "future"**: The Rust language has been evolving since its inception, and we expect it to continue to evolve going forward.
+These evolutionary steps represent a traversal over the language design landscape.
+For every Rust release, we expect the current implementation to be standing at one point in that landscape, while the Rust community's idealized goal awaits at some higher point up the mountain.
+
+**"prescriptive", "descriptive"**: A descriptive dictionary is one that attempts to describe how a word *is* used, while a prescriptive dictionary is one that prescribes how a word *should be* used.
+We take inspiration from that distinction to tease apart two important audience types.
+
+Rust provides a stability promise: "You should never have to fear upgrading to a new version of stable Rust."
+That raises a natural question: Why does the purpose statement distinguish version-crossing prescriptive definitions from version-specific descriptive definitions?
+
+Our answer:
+That stability promise left a bit of wiggle-room for itself, in terms of what the project considers "fear" vs "reasonable labor associated with a Rust upgrade."
+When defining semantics, one must be more explicit about any such wiggle-room.
+Some Rust users *need* a description of the expected semantics as it stands for the Rust release that sits in their hands; they are the audience for version-specific details.
+But other Rust users, such as some library developers, have a more forward-looking perspective.
+The forward-looking developers may require an assurance that one specific code snippet A will always be accepted, and will also always have a particular meaning.
+They may require an assurance that a different snippet B will *never* be accepted.
+Or they may require an assurance that a third snippet C leveraging Unsafe Rust will always have undefined behavior (e.g. to justify a local transformation by arguing that no *new* undefined behavior is injected by that transformation.)
+These are all cases that call for a prescriptive definition of the semantics; it does not matter in these cases what the compiler currently does -- what matters is what it will do in the future, which is inherently prescriptive.
+
+It would be premature to fix firm definitions in those areas, e.g. categorizing for each input program whether it is accepted or rejected by the type inference system, and then forcing all futures versions of Rust to follow that same categorization.
```

Comment:
Is this supposed to be something like "There are other areas where it would be premature to fix firm definitions"? Otherwise this sounds like it's contradicting the previous paragraph.

Reply:
I think the text says what it means.

It is not meant to be contradicting the previous paragraph.

The previous paragraph is saying that some users *need* prescriptive definitions; that detailed statements that only speak about the current version of Rust will not serve their purposes.

This paragraph is saying that the we need to take care in what the prescriptive definitions say, and not prematurely commit to things that will have to be overturned. So we explicitly frame things in terms of *bounds* on behavior.

Reply:
You're right. I'm going to try to incorporate your feedback now.

Reply:
Hopefully addressed by https://github.com/rust-lang/blog.rust-lang.org/pull/1166/commits/f72debebb2e697554b4b77e15e6d45aa37cbbc83